### PR TITLE
Use user-agent instead of screen_width to get device type

### PR DESCRIPTION
**Pull request by ukutaht**
### Changes

Follow-up for https://github.com/plausible/analytics/pull/1659:
* More exhaustive matching for device types from matomo/device-detector
* Reverted a change to the UI labels for now. Not sure if that's desirable.

### Tests
- [x] Automated tests have been added

### Changelog
- [ ] Entry has been added to changelog

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

File: lib/plausible/ingestion/event.ex [hunk 1]
```diff
@@ -307,6 +294,24 @@ defmodule Plausible.Ingestion.Event do
...
+      %UAInspector.Result.Device{type: "portable media player"} -> "Mobile"
+      %UAInspector.Result.Device{type: "phablet"} -> "Mobile"
+      %UAInspector.Result.Device{type: "wearable"} -> "Mobile"
+      %UAInspector.Result.Device{type: "camera"} -> "Mobile"
+      %UAInspector.Result.Device{type: "car browser"} -> "Tablet"
+      %UAInspector.Result.Device{type: "tablet"} -> "Tablet"
+      %UAInspector.Result.Device{type: "tv"} -> "Desktop"
+      %UAInspector.Result.Device{type: "console"} -> "Desktop"
+      %UAInspector.Result.Device{type: "desktop"} -> "Desktop"
+      _ -> nil
```
Comment (vinibrsl) [inline lib/plausible/ingestion/event.ex:311]: I guess you covered most device types if not all. Is it worth logging these `nil` cases somewhere so we can keep track of what types we're not handling?

**Review by vinibrsl — APPROVED** [review]
Looks good! I left a minor comment :)

File: lib/plausible/ingestion/event.ex (hunk 1 above)
Comment (ukutaht) [reply lib/plausible/ingestion/event.ex:311]: Yes, good call. I've added a `Sentry.capture_message` when we get a result from UAInspector that we don't know how to handle.

**Comment by ukutaht** [comment]
I tried to mask the historical 'Laptop' entry as well. The DB query is pretty simple:

```sql
SELECT if(screen_size = 'Laptop', 'Desktop', screen_size) AS device FROM sessions WHERE ... GROUP BY device
```

Unfortunately I cannot find a way to make this work with Ecto. I need to be able to refer to the alias device in GROUP BY but I can't seem to find a way to make that work.

**Comment by ruslandoga** [comment]
@ukutaht `selected_as` might work.

**Comment by ukutaht** [comment]
> @ukutaht [`selected_as`](https://hexdocs.pm/ecto/Ecto.Query.API.html#selected_as/2) might work. It's not supported in the current adapter, but it is supported in `chto` :)

Oh cool, I didn't know about `selected_as`. Too bad the current adapter doesn't support it. I can maybe revisit this once the `chto` adapter is production-ready :)

**Comment by aerosol** [comment]
Hacky but this works:

@ukutaht 

```elixir
from(s in "sessions", 
  select: %{
    device: fragment("if(? = 'Laptop', 'Desktop', ?) as device", s.screen_size, s.screen_size),
    count: sum(s.sign)
  },
  group_by: fragment("device")
) 
```

```sql
SELECT if(s0."screen_size" = 'Laptop', 'Desktop', s0."screen_size") as device, sum(s0."sign") FROM "sessions" AS s0 GROUP BY device []
```

```elixir
[
  %{count: 2623, device: ""},
  %{count: 45948, device: "Mobile"},
  %{count: 46189, device: "Tablet"},
  %{count: 91823, device: "Desktop"}
]
```

**Comment by ukutaht** [comment]
> Hacky but this works:
> 
> @ukutaht
> 
> ```elixir
> from(s in "sessions", 
>   select: %{
>     device: fragment("if(? = 'Laptop', 'Desktop', ?) as device", s.screen_size, s.screen_size),
>     count: sum(s.sign)
>   },
>   group_by: fragment("device")
> ) 
> ```
> 
>  ```sql
> SELECT if(s0."screen_size" = 'Laptop', 'Desktop', s0."screen_size") as device, sum(s0."sign") FROM "sessions" AS s0 GROUP BY device []
> ```
> 
>  ```elixir
> [
>   %{count: 2623, device: ""},
>   %{count: 45948, device: "Mobile"},
>   %{count: 46189, device: "Tablet"},
>   %{count: 91823, device: "Desktop"}
> ]
> ```

Here's my attempt: https://github.com/plausible/analytics/commit/1ebe5b09b07643624960f0b3d2240e912dfac6f0

It works in some cases but not all. Some tests are failing.

**Comment by aerosol** [comment]
@ukutaht no `sign` clause necessary?

**Comment by ukutaht** [comment]
> @ukutaht no `sign` clause necessary?

Are you asking why my attempt doesn't select `sign`? It's because this query is composed in multiple places and the SELECT clauses for metrics are in `stats/clickhouse/base.ex`

**Comment by ukutaht** [comment]
Closed in favour of https://github.com/plausible/analytics/pull/2680

**Pull request closed without merging**